controller/sync_namespace.go

```diff
@@ -0,0 +1,49 @@
+package controller
+
+import (
+	"fmt"
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/argo"
+	gitopscommon "github.com/argoproj/gitops-engine/pkg/sync/common"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func syncNamespace(resourceTracking argo.ResourceTracking, appLabelKey string, trackingMethod v1alpha1.TrackingMethod, appName string, syncPolicy *v1alpha1.SyncPolicy) func(un *unstructured.Unstructured) (bool, error) {
+	return func(un *unstructured.Unstructured) (bool, error) {
+		isNewNamespace := un != nil && un.GetUID() == "" && un.GetResourceVersion() == ""
+
+		if un != nil && syncPolicy != nil {
+			// managedNamespaceMetadata relies on SSA, and since the diffs are computed by the k8s control plane we
+			// always need to call the k8s api server, so we'll always need to return true if managedNamespaceMetadata is set.
+			hasManagedMetadata := syncPolicy.ManagedNamespaceMetadata != nil
+			if hasManagedMetadata {
+				managedNamespaceMetadata := syncPolicy.ManagedNamespaceMetadata
+				un.SetLabels(managedNamespaceMetadata.Labels)
+				un.SetAnnotations(appendNamespaceSSA(managedNamespaceMetadata.Annotations))
+
+				// TODO: This should probably be removed from this block and be tracked for all namespaces created by ArgoCD.
+				// However, there's an issue with a failing integration test (see app_management_test#TestNamespaceAutoCreation)
+				shouldTrackNs := resourceTracking.GetAppInstance(un, appLabelKey, trackingMethod) == nil && appLabelKey != "" && appName != ""
```

Comment:
The setting of the resource tracking was originally outside the `hasManagedMetadata` block, but having it there causes an E2E test to fail (see `TestNamespaceAutoCreation` in `app_management_test.go`). I'm not sure if this is a legitimate failure or if this is an issue with the test itself. 

On that test, setting an app namespace makes it pass, otherwise it fails..

Apart from getting to the bottom of why that test would be failing, the question is whether it makes sense to only set resource tracking if `managedNamespaceMetadata` is set, or if we should try to also set resource tracking for new namespaces as well, regardless if `managedNamespaceMetadata` is set or not? WDYT @crenshaw-dev @leoluz?

Reply:
I think we should keep the previous behaviour unless we have a good reason to change it. Was the create namespace feature setting the resource tracking? If not, then I think we should keep it this way.